From 789d6703b1372ec59ddf65a7bc87bf28f9186acd Mon Sep 17 00:00:00 2001
From: Xianzhong <b07117@freescale.com>
Date: Fri, 1 Aug 2014 18:44:20 +0800
Subject: [PATCH] ENGR00325794 [#1087] fix video memory mutex sharing issue

the root cause is video memory mutex is not global variable,
it will cause video memory managment problem with mixed 2D/3D/VG.

kernel panic with multiple instances stress test running glesx_viv.sh.

Date: Jul 31, 2014
Signed-off-by: Xianzhong <b07117@freescale.com>
Acked-by: Jason Liu
(cherry picked from commit 9cec1cbd7ca2378e5c429f57d088d23d73d9c2f3)
---
 drivers/mxc/gpu-viv/hal/kernel/gc_hal_kernel.c     |    4 +---
 drivers/mxc/gpu-viv/hal/kernel/inc/gc_hal.h        |    7 +++++++
 .../hal/os/linux/kernel/gc_hal_kernel_linux.h      |    2 ++
 .../gpu-viv/hal/os/linux/kernel/gc_hal_kernel_os.c |   19 +++++++++++++++++++
 4 files changed, 29 insertions(+), 3 deletions(-)

diff --git a/drivers/mxc/gpu-viv/hal/kernel/gc_hal_kernel.c b/drivers/mxc/gpu-viv/hal/kernel/gc_hal_kernel.c
index d24a08b..9b558ef 100644
--- a/drivers/mxc/gpu-viv/hal/kernel/gc_hal_kernel.c
+++ b/drivers/mxc/gpu-viv/hal/kernel/gc_hal_kernel.c
@@ -405,7 +405,7 @@ gckKERNEL_Construct(
     gcmkONERROR(gckKERNEL_SecurityOpen(kernel, kernel->core, &kernel->securityChannel));
 #endif
     /* Construct a video memory mutex. */
-    gcmkONERROR(gckOS_CreateMutex(Os, &kernel->vidmemMutex));
+    gcmkONERROR(gckOS_GetVideoMemoryMutex(Os, &kernel->vidmemMutex));
 
     /* Return pointer to the gckKERNEL object. */
     *Kernel = kernel;
@@ -634,8 +634,6 @@ gckKERNEL_Destroy(
     gcmkVERIFY_OK(gckKERNEL_SecurityClose(Kernel->securityChannel));
 #endif
 
-    gcmkVERIFY_OK(gckOS_DeleteMutex(Kernel->os, Kernel->vidmemMutex));
-
     /* Mark the gckKERNEL object as unknown. */
     Kernel->object.type = gcvOBJ_UNKNOWN;
 
diff --git a/drivers/mxc/gpu-viv/hal/kernel/inc/gc_hal.h b/drivers/mxc/gpu-viv/hal/kernel/inc/gc_hal.h
index 02d0014..b4743b3 100644
--- a/drivers/mxc/gpu-viv/hal/kernel/inc/gc_hal.h
+++ b/drivers/mxc/gpu-viv/hal/kernel/inc/gc_hal.h
@@ -1560,6 +1560,13 @@ gckOS_StopTimer(
     IN gctPOINTER Timer
     );
 
+/* Get the global video memory mutex. */
+gceSTATUS
+gckOS_GetVideoMemoryMutex(
+    IN gckOS Os,
+    OUT gctPOINTER *Mutex
+    );
+
 /******************************************************************************\
 ********************************* gckHEAP Object ********************************
 \******************************************************************************/
diff --git a/drivers/mxc/gpu-viv/hal/os/linux/kernel/gc_hal_kernel_linux.h b/drivers/mxc/gpu-viv/hal/os/linux/kernel/gc_hal_kernel_linux.h
index 88fbf7d..1dc29a3 100644
--- a/drivers/mxc/gpu-viv/hal/os/linux/kernel/gc_hal_kernel_os.c
+++ b/drivers/mxc/gpu-viv/hal/os/linux/kernel/gc_hal_kernel_os.c
@@ -222,6 +222,8 @@ struct _gckOS
 
     int                         gpu_clk_on[3];
     struct mutex                gpu_clk_mutex;
+
+    gctPOINTER                  vidmemMutex;
 };
 
 typedef struct _gcsSIGNAL * gcsSIGNAL_PTR;
diff --git a/drivers/mxc/gpu-viv/hal/os/linux/kernel/gc_hal_kernel_os.c b/drivers/mxc/gpu-viv/hal/os/linux/kernel/gc_hal_kernel_os.c
index 9184e9b..89fe678 100644
--- a/drivers/mxc/gpu-viv/hal/os/linux/kernel/gc_hal_kernel_os.c
+++ b/drivers/mxc/gpu-viv/hal/os/linux/kernel/gc_hal_kernel_os.c
@@ -917,6 +917,9 @@ gckOS_Construct(
 
     gckOS_ImportAllocators(os);
 
+    /* Construct a video memory mutex. */
+    gcmkONERROR(gckOS_CreateMutex(os, &os->vidmemMutex));
+
     /* Return pointer to the gckOS object. */
     *Os = os;
 
@@ -1029,6 +1032,9 @@ gckOS_Destroy(
     /* Destroy debug lock mutex. */
     gcmkVERIFY_OK(gckOS_DeleteMutex(Os, Os->debugLock));
 
+    /* Destroy video memory mutex. */
+    gcmkVERIFY_OK(gckOS_DeleteMutex(Os, Os->vidmemMutex));
+
     /* Wait for all works done. */
     flush_workqueue(Os->workqueue);
 
@@ -8723,6 +8723,20 @@ gckOS_GetProcessNameByPid(
     return gcvSTATUS_OK;
 }
 
+gceSTATUS
+gckOS_GetVideoMemoryMutex(
+    IN gckOS Os,
+    OUT gctPOINTER *Mutex
+    )
+{
+    gcmkHEADER_ARG("Mutex=x%X", Mutex);
+
+    *Mutex = Os->vidmemMutex;
+
+    gcmkFOOTER_NO();
+    return gcvSTATUS_OK;
+}
+
 #if gcdANDROID_NATIVE_FENCE_SYNC
 
 gceSTATUS
-- 
1.7.1

